python/ray/data/tests/test_auto_parallelism.py

```diff
@@ -1,10 +1,106 @@
 import pytest
+from dataclasses import dataclass
 
 import ray
 from ray.data.context import DatasetContext
+from ray.data._internal.util import _autodetect_parallelism
 from ray.tests.conftest import *  # noqa
 
 
+@dataclass
+class TestCase:
+    avail_cpus: int
+    data_size: int
+    expected_parallelism: int
+
+
+MB = 1024 * 1024
+GB = 1024 * MB
+
+TEST_CASES = [
+    TestCase(
+        avail_cpus=4,
+        data_size=1024,
+        expected_parallelism=8,  # avail_cpus has precedence
+    ),
+    TestCase(
+        avail_cpus=4,
+        data_size=1 * MB,
+        expected_parallelism=10,  # MIN_BLOCK_SIZE has precedence
+    ),
+    TestCase(
+        avail_cpus=4,
+        data_size=2 * MB,
+        expected_parallelism=20,  # MIN_BLOCK_SIZE has precedence
+    ),
+    TestCase(
+        avail_cpus=4,
+        data_size=10 * MB,
+        expected_parallelism=102,  # MIN_BLOCK_SIZE has precedence
+    ),
+    TestCase(
+        avail_cpus=4,
+        data_size=1 * GB,
+        expected_parallelism=200,  # MIN_PARALLELISM has precedence
+    ),
+    TestCase(
+        avail_cpus=4,
+        data_size=10 * GB,
+        expected_parallelism=200,  # MIN_PARALLELISM has precedence
+    ),
+    TestCase(
+        avail_cpus=150,
+        data_size=10 * GB,
+        expected_parallelism=300,  # avail_cpus has precedence
+    ),
+    TestCase(
+        avail_cpus=400,
+        data_size=10 * GB,
+        expected_parallelism=800,  # avail_cpus has precedence
+    ),
+    TestCase(
+        avail_cpus=400,
+        data_size=1 * MB,
+        expected_parallelism=800,  # avail_cpus has precedence
+    ),
+    TestCase(
+        avail_cpus=4,
+        data_size=1000 * GB,
+        expected_parallelism=2000,  # MAX_BLOCK_SIZE has precedence
+    ),
+    TestCase(
+        avail_cpus=4,
+        data_size=10000 * GB,
+        expected_parallelism=20000,  # MAX_BLOCK_SIZE has precedence
+    ),
+]
+
+
+def run_test(test: TestCase):
+    print("TESTING", test)
+
+    class MockReader:
+        def __init__(self, size):
+            self.size = size
+
+        def estimate_inmemory_data_size(self):
+            return self.size
+
+    result, _ = _autodetect_parallelism(
+        parallelism=-1,
+        cur_pg=None,
+        ctx=DatasetContext.get_current(),
+        reader=MockReader(test.data_size),
+        avail_cpus=test.avail_cpus,
+    )
+    assert result == test.expected_parallelism, (result, test)
+
+
+def test_autodetect_parallelism():
+    for test in TEST_CASES:
+        run_test(test)
```

Comment:
Could you instead run these test cases via parametrizing the test, so there's one test run per test case? In addition to that being more idiomatic pytest test design, if a single test case fails, it will be a lot easier to rerun that specific test.

Reply:
Hmm, it seems kind of sketchy to parameterize with an TestCase object, it's not clear to me pytest param is good for table-driven tests like this. I also think this way is more straightforward / simpler if you're not a pytest expert.

Reply:
>it's not clear to me pytest param is good for table-driven tests like this

It's definitely built for patterns exactly like this:
```python
@pytest.mark.parametrize(
    "avail_cpus,data_size,expected",
    [
        (4, 1024, 8),
        (4, 1 * MB, 10),
        (4, 2 * MB, 20),
    ],
)
def test_autodetect_parallelism(avail_cpus, data_size, expected):
    print("TESTING", test)

    class MockReader:
        def __init__(self, size):
            self.size = size

        def estimate_inmemory_data_size(self):
            return self.size

    result, _ = _autodetect_parallelism(
        parallelism=-1,
        cur_pg=None,
        ctx=DatasetContext.get_current(),
        reader=MockReader(data_size),
        avail_cpus=avail_cpus,
    )
    assert result == expected, (result, test)
```

>I also think this way is more straightforward / simpler if you're not a pytest expert.

I disagree a bit there, this is one of the core features of pytest, and is a big dev UX improvement. pytest is our core testing library for Python code so I think asking devs to use this basic pytest feature when implementing such a test case pattern is reasonable.

IMO we should really be trying to move our unit test suites towards a consistent arrange-act-assert pattern for each individual test instead of having these big mono-tests, and this is the idiomatic pytest way to run multiple test cases through a single arrange-act-assert test.

Reply:
I tried this out, but it's considerably less readable. When scanning the table, you just can't really tell what's going on since there aren't any kwargs for the test table.

If you can generate pytest cases given dataclasses, that would be a good alternative that wouldn't sacrifice the readability.

Reply:
Update: turns out astuple() is pretty convenient, changed to use that.